sherpa/models/template.py

```diff
@@ -26,149 +84,336 @@
-                   at a specific parameter vector (corresponds to a row
-                   in `parvals`).
+    if parvals.ndim != 2:
+        raise ValueError(f"parvals must be 2D, sent {parvals.ndim}D")
 
-    `template_interpolator_name` - name of the template interpolator, or None
-                   for disabling interpolation *between* templates.
-                   See load_template_model for more information.
+    # It is not obvious why we have the data transposed, but leave as
+    # is for now.
+    #
+    if parvals.shape[1] != len(names):
+        raise ValueError(f"number of parvals and names do not match: {parvals.shape[1]} vs {len(names)}")
 
-    """
     # Create a list of parameters from input
     pars = []
-    for ii, name in enumerate(names):
-        minimum = min(parvals[:, ii])
-        maximum = max(parvals[:, ii])
-        initial = parvals[:, ii][0]
+    for name, parval in zip(names, parvals.T):
+        minimum = min(parval)
+        maximum = max(parval)
         # Initial parameter value is always first parameter value listed
-        par = Parameter(modelname, name, initial,
-                        minimum, maximum,
-                        minimum, maximum)
+        par = Parameter(modelname, name, parval[0],
+                        min=minimum, max=maximum,
+                        hard_min=minimum, hard_max=maximum)
         pars.append(par)
 
     # Create the templates table from input
     tm = TemplateModel(modelname, pars, parvals, templates)
-    if template_interpolator_name is not None:
-        if template_interpolator_name in interpolators:
-            interp = interpolators[template_interpolator_name]
-            args = interp[1]
-            args['template_model'] = tm
-            args['name'] = modelname
-            return interp[0](**args)
-    else:
+    if template_interpolator_name is None:
         return tm
 
+    try:
+        func, args = interpolators[template_interpolator_name]
+    except KeyError:
+        raise ModelErr(f"Unknown template_interpolator_name={template_interpolator_name}") from None
+
+    # Do not add keys to the stored dictionary
+    args = args.copy()
+    args['template_model'] = tm
+    args['name'] = modelname
+    return func(**args)
+
 
+# Should this provide direct access to query to better match
+# TemplateModel.  Or should we not export the parvals attribute, as we
+# don't want to copy that interface?
+#
 class InterpolatingTemplateModel(ArithmeticModel):
+    """Allow parameter interpolation for a TemplateModel.
+
+    This class should not be used directly.
+
+    Parameters
+    ----------
+    name : str
+        The name of the model
+    template_model : TemplateModel instance
+        The templates to use.
+
+    See Also
+    --------
+    TemplateModel, KNNInterpolator
+
+    """
+
     def __init__(self, name, template_model):
         self.template_model = template_model
+        self.parvals = template_model.parvals
```

Comment:
Note to self: can `parvals` ever differ from `template_model.parvals`? If so, do we care?

Reply:
I'm just not sure we care about this setting, but I don't want to change it as maybe there's some reason for it!

Reply:
There are instances when using the exact parameters and not interpolated values is useful. Basically for large number of templates with many parameters the initial search via gridsearch with no interpolation would provide some idea of restricting the fitting with interpolation. I guess originally this was helpful in debugging/testing the code.